lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,63 +1175,186 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
+    SwiftAutoLinkExtract,
+  };
+
+  ValueTy Value;
+
+  AutolinkKind(ValueTy value) : Value(value) {}
+  AutolinkKind(const AutolinkKind &kind) : Value(kind.Value) {}
+  operator enum ValueTy() { return Value; }
+
+  StringRef getSectionNameMetadata() {
+    // FIXME: This constant should be vended by LLVM somewhere.
+    switch (Value) {
+    case AutolinkKind::LLVMDependentLibraries:
+      return "llvm.dependent-libraries";
+      break;
+    case AutolinkKind::LLVMLinkerOptions:
+    case AutolinkKind::SwiftAutoLinkExtract:
+      return "llvm.linker.options";
+      break;
+    }
+  }
+
+  void collectEntriesFromLibraries(
+      llvm::SmallVector<llvm::MDNode *, 4> &Entries,
+      const SmallVector<LinkLibrary, 32> &AutolinkEntries, IRGenModule &IGM) {
+    llvm::LLVMContext &ctx = IGM.getLLVMContext();
+
+    switch (Value) {
+    case AutolinkKind::LLVMLinkerOptions:
+    case AutolinkKind::SwiftAutoLinkExtract: {
+      // On platforms that support autolinking, continue to use the metadata.
+      for (LinkLibrary linkLib : AutolinkEntries) {
+        switch (linkLib.getKind()) {
+        case LibraryKind::Library: {
+          llvm::SmallString<32> opt =
+              getTargetDependentLibraryOption(IGM.Triple, linkLib.getName());
+          Entries.push_back(
+              llvm::MDNode::get(ctx, llvm::MDString::get(ctx, opt)));
+          break;
+        }
+        case LibraryKind::Framework: {
+          llvm::Metadata *args[] = {
+              llvm::MDString::get(ctx, "-framework"),
+              llvm::MDString::get(ctx, linkLib.getName())};
+          Entries.push_back(llvm::MDNode::get(ctx, args));
+          break;
+        }
+        }
+      }
+      break;
+    }
+    case AutolinkKind::LLVMDependentLibraries: {
+      for (LinkLibrary linkLib : AutolinkEntries) {
+        switch (linkLib.getKind()) {
+        case LibraryKind::Library: {
+          Entries.push_back(llvm::MDNode::get(
+              ctx, llvm::MDString::get(ctx, linkLib.getName())));
+          break;
+        }
+        case LibraryKind::Framework: {
+          llvm_unreachable(
+              "llvm.dependent-libraries doesn't support framework dependency");
+          break;
+        }
+        }
+      }
+      break;
+    }
+    }
+  }
+
+  void writeEntries(const llvm::SmallVector<llvm::MDNode *, 4> &Entries,
+                    llvm::NamedMDNode *Metadata, IRGenModule &IGM) {
+    switch (Value) {
+    case AutolinkKind::LLVMLinkerOptions:
+    case AutolinkKind::LLVMDependentLibraries: {
+      // On platforms that support autolinking, continue to use the metadata.
+      Metadata->clearOperands();
+      for (auto *Entry : Entries)
+        Metadata->addOperand(Entry);
+      break;
+    }
+    case AutolinkKind::SwiftAutoLinkExtract: {
+      // Merge the entries into null-separated string.
+      llvm::SmallString<64> EntriesString;
+      for (auto EntryNode : Entries) {
+        const llvm::MDNode *MD = cast<llvm::MDNode>(EntryNode);
```

Comment:
just an FYI, if the type is already on the same line, please use auto. But please make sure to include the const and the *, so:

```
const auto *MD = cast<llvm::MDNode>(...)
```